internal/sync/syncer.go

```diff
@@ -0,0 +1,311 @@
+			keyCh <- key
+		}
+	}
+
+	// Following for loop keeps calling assign() function to assign watching tasks to the workers.
+	// It makes sure each element in the list will be assigned every N milliseconds.
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Shutting down synchronizer job assigner.")
+			return nil
+		default:
+			assign()
+		}
+		// Make sure each of the key will be assigned at most every N milliseconds.
+		time.Sleep(time.Millisecond * time.Duration(func() int {
+			l := s.Length()
+			if l == 0 {
+				return s.options.taskInterval
+			}
+			result := s.options.taskInterval / l
+			if result > 0 {
+				return result
+			}
+			return 1
+		}()))
+	}
+}
+
+// Function run() defines each of the worker's job.
+// It waits for keys in the channel, and starts a synchronization job
+func (s *Syncer) run(ctx context.Context, id int, keyCh <-chan string) {
+	log := logging.FromContext(ctx)
+	log.Infof("Started synchronizer worker %v", id)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("Stopped synchronizer worker %v", id)
+			return
+		case key := <-keyCh:
+			if err := s.runOnce(ctx, key, id); err != nil {
+				log.Errorw("Failed to execute a task", zap.String("gitSyncKey", key), zap.Error(err))
+			}
+		}
+	}
+}
+
+// Function runOnce implements the logic of each synchronization.
+func (s *Syncer) runOnce(ctx context.Context, key string, worker int) error {
+	log := logging.FromContext(ctx).With("worker", fmt.Sprint(worker)).With("gitSyncKey", key)
+	log.Debugf("Working on key: %s.", key)
+	strs := strings.Split(key, "/")
+	if len(strs) != 2 {
+		return fmt.Errorf("invalid key %q", key)
+	}
+	namespace := strs[0]
+	gitSyncName := strs[1]
+	gitSync := &v1alpha1.GitSync{}
+	if err := s.client.Get(ctx, client.ObjectKey{Namespace: namespace, Name: gitSyncName}, gitSync); err != nil {
+		if apierrors.IsNotFound(err) {
+			s.StopWatching(key)
+			log.Info("No corresponding GitSync found, stopped watching.")
+			return nil
+		}
+		return fmt.Errorf("failed to query GitSync object of key %q, %w", key, err)
+	}
+	if !gitSync.GetDeletionTimestamp().IsZero() {
+		s.StopWatching(key)
+		log.Debug("GitSync object being deleted.")
+		return nil
+	}
+
+	globalConfig, err := controllerConfig.GetConfigManagerInstance().GetConfig()
+	if err != nil {
+		log.Errorw("error getting  the  global config", "err", err)
+	}
+	repo, err := git.CloneRepo(ctx, s.client, gitSync, namespace, globalConfig)
+	if err != nil {
+		return fmt.Errorf("failed to clone the repo of key %q, %w", key, err)
+	}
+	manifests, err := git.GetLatestManifests(ctx, repo, gitSync)
+	if err != nil {
+		return fmt.Errorf("failed to get the manifest of key %q, %w", key, err)
+	}
+	uns, err := toUnstructuredAndApplyAnnotation(manifests, gitSyncName)
+	if err != nil {
+		return fmt.Errorf("failed to parse the manifest of key %q, %w", key, err)
+	}
+	synced := s.sync(gitSync, uns)
+	if synced {
+		log.Info("GitSync object is successfully synced.")
+	}
```

Comment:
Does the gitops-engine assume that the full set of manifests passed here represent the complete set of Resources associated with the GitSync? Is that how it know that a manifest has been deleted?